Optimize backward for torch.repeat

ejguan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46726 Optimize backward for torch.repeat**

Fixes #43192

Differential Revision: [D24739840](https://our.internmc.facebook.com/intern/diff/D24739840)

ejguan committed:
[WIP] Optimize backward for torch.repeat

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -636,13 +636,23 @@ Tensor repeat_backward(Tensor grad, IntArrayRef repeats, IntArrayRef input_shape
   }
-  for (size_t j = num_unsqueezed; j < repeats.size(); ++j) {
+
+  const auto grad_sizes = grad.sizes();
```
nit: can this be taken by reference?

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -636,13 +636,23 @@ Tensor repeat_backward(Tensor grad, IntArrayRef repeats, IntArrayRef input_shape
     int64_t dim = j - num_unsqueezed;
-    grad = sum_tensorlist(grad.chunk(repeat, dim));
+
+    grad_iter_ = grad_sizes_.insert(grad_iter_, repeat);
```
Maybe add a comment here to mention what the algo does and why the new shape should be `(repeat, dimsize/repeat)` (and not `(dimsize/repeat, repeat)`).

albanD reviewed (commented):
Change looks good. Just small comments.

ejguan commented:
Tests:
```python
x = torch.randn(3, 4, requires_grad=True)
def test_repeat(x):
    y = x.repeat(20, 10, 15, 25)
    out = y.sum()
    out.backward()
```
Average time for 10 * 10000 runs.
|     	        | Previous 	| Current 	| Improvement |
|:---:	|:--------:	|:-------:	|:-------:	         |
| CPU 	|        4.236 ms  	|        4.124 ms 	|       2.64%  	|
| GPU 	|        760.262 us  	|        259.652 us 	|       65.85%  	|

ejguan commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -636,13 +636,23 @@ Tensor repeat_backward(Tensor grad, IntArrayRef repeats, IntArrayRef input_shape
       continue;
     }
     int64_t dim = j - num_unsqueezed;
-    grad = sum_tensorlist(grad.chunk(repeat, dim));
```
Thank you!

ejguan committed:
Update on "[WIP] Optimize backward for torch.repeat"


Fixes #43192

Differential Revision: [D24481801](https://our.internmc.facebook.com/intern/diff/D24481801)

[ghstack-poisoned]

zou3519 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -615,17 +615,6 @@ Tensor renorm_backward(const Tensor & grad, const Tensor & self, Scalar p, int64
   return at::where(mask, grad, grad_norm);
 }
 
-Tensor sum_tensorlist(TensorList tl) {
```
We should delete sum_tensorlist from FunctionsManual.h

zou3519 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -636,13 +625,57 @@ Tensor repeat_backward(Tensor grad, IntArrayRef repeats, IntArrayRef input_shape
+    //  [g1_0+g2_0, g1_1+g2_1]]          [[g2_0, g2_1],             [g2_2, g2_3],
+    //                                    [g2_2, g2_3],             [g2_4, g2_5]]
+    //                                    [g2_4, g2_5]]]
+    // If gradient tensor is reshaped to [..., dimsize/repeat, repeat, ...] and then
+    // sum over 'dim+1'. The gradient for input is not correctly aligned with . 
```
"The gradient for input is not correctly aligned with . " -> what is the " . " ?

ejguan replied:
It should be "If gradient tensor is reshaped to [..., dimsize/repeat, repeat, ...] and then sum over 'dim+1'. The gradient for input is not correctly aligned with input"

zou3519 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -636,13 +625,57 @@ Tensor repeat_backward(Tensor grad, IntArrayRef repeats, IntArrayRef input_shape
-  for (size_t j = num_unsqueezed; j < repeats.size(); ++j) {
+
+  const auto & grad_sizes = grad.sizes();
+  std::vector<int64_t> grad_sizes_(input_dims);
```
We prefer DimVector to `std::vector<int64_t>` most of the time. https://github.com/pytorch/pytorch/blob/daf2a6a29da5109157bb44fc5e4dbc4fb5c56c15/aten/src/ATen/core/DimVector.h#L11.

The first 5 elements of DimVector go on the stack and the rest go onto the heap. c.f. all elements of std::vector go on the heap. DimVector improves performance for keeping track of metadata that has 5 or fewer elements, which is the case for most PyTorch use cases out in the wild

zou3519 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -636,13 +625,57 @@ Tensor repeat_backward(Tensor grad, IntArrayRef repeats, IntArrayRef input_shape
     }
     int64_t dim = j - num_unsqueezed;
-    grad = sum_tensorlist(grad.chunk(repeat, dim));
+    // Reshape gradient
+    // Index:      [..., dim,   , ...]    [..., dim,    dim+1,        , ...]
+    // Shape: From [..., dimsize, ...] to [..., repeat, dimsize/repeat, ...]
+    // The gradient tensor at 'dim' is reshaped to 'repeat' times of input tensor. 
+    // Then, sum up gradients over repeated tensors along 'dim', and reduce shape 
+    // from 'repeat * dimsize/repeat' to 'dimsize/repeat' ('input_dimsize').
```
This comment is beautiful

zou3519 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -636,13 +625,57 @@ Tensor repeat_backward(Tensor grad, IntArrayRef repeats, IntArrayRef input_shape
+  std::vector<int64_t> grad_sizes_(input_dims);
+  std::copy(grad_sizes.begin(), grad_sizes.end(), grad_sizes_.begin());
+  auto grad_iter_ = grad_sizes_.begin();
+  for (size_t j = num_unsqueezed; j < repeats.size(); ++j, ++grad_iter_) {
```
(no action required)

I think we can call (reshape + sum) a single time instead multiple times. This may be more efficient (because it fuses operations together, especially the sum) but I'm not completely sure. Here's the idea:

Let the input sizes be `input_dimsize`, the number of repeats be `repeats`, and assume that the grad has the same dimensionality as the input.

Then: we could just reshape the grad in one call to be:
```
grad_reshaped = grad.reshape([ repeats[0], input_dimsize[0] / repeats[0], repeats[1], input_dimsize[1] / repeats[1], repeats[2], input_dimsize[2] / repeats[2], ...])
``` 
And reduce over all of the repeat dimensions with:
```
grad_reshape.sum([0, 2, 4, 6, 8, ...])
```

That being said, the code written as-is already delivers a nice improvement so I think it's fine if we ship it in the current form and worry about optimizing it more in the future

ejguan replied:
I actually was also thinking this implementation. But, in the other hand, we need to do `reshape` and `sum` on several dimension where `repeat == 1`. I should also take them into my test cases to compare the performance between two implementations.

zou3519 replied:
> But, in the other hand, we need to do reshape and sum on several dimension where repeat == 1

Yeah, that's the downside so it is hard to gauge whether or not it would actually be faster. One would expect a sum on a dimension of size 1 to be a no-op (or close to one) and but the reshape might cause a copy to happen when it didn't happen before (I'm not sure about this part, would need to think it through with some examples)

zou3519 replied:
If you're interested in testing the performance between these two implementations it would be good to try some non-contiguous tensors as well

zou3519 commented:
> Tests (using `torch.utils.benchmark`):

I'm curious to see the equivalent torch.expand / torch.repeat_interleave call for  the examples in https://github.com/pytorch/pytorch/issues/43192#issuecomment-712262196.

Also, it would be nice to see some benchmark figures on larger tensors (a tensor of size [3, 4] is pretty small, in practice users have larger tensors)

zou3519 reviewed (dismissed):
this lgtm (with some minor comments). I'm curious to see some more performance numbers

ejguan committed:
Update on "Optimize backward for torch.repeat"


Fixes #43192

Differential Revision: [D24481801](https://our.internmc.facebook.com/intern/diff/D24481801)

[ghstack-poisoned]

ejguan committed:
Update on "Optimize backward for torch.repeat"


Fixes #43192

Differential Revision: [D24481801](https://our.internmc.facebook.com/intern/diff/D24481801)

[ghstack-poisoned]

ejguan commented:
Benchmark for two different implementations.
Test 1:
```python
x = torch.randn(320, 480, requires_grad=True)
x_t = torch.randn(320, 480, requires_grad=True).t() # non-contiguous
...
y = x.repeat(16, 16)
...
```
Test 2:
```python
x = torch.randn(160, 240, requires_grad=True)
x_t = torch.randn(160, 240, requires_grad=True).t() # non-contiguous
...
y = x.repeat(32, 32)
...
```
Test 3:
```python
x = torch.randn(80, 120, requires_grad=True)
x_t = torch.randn(80, 120, requires_grad=True).t() # non-contiguous
...
y = x.repeat(64, 64)
...
```
Average time for 1,000 runs:
|                 	|    CPU    	|   GPU   	|   CPU NC  	|  GPU NC 	|
|:---------------:	|:---------:	|:-------:	|:---------:	|:-------:	|
|    Before (1)   	| 125.31 ms 	| 4.46 ms 	| 155.39 ms 	| 4.72 ms 	|
|  Multi-time (1) 	| 127.16 ms 	| 4.02 ms 	| 165.37 ms 	| 4.14 ms 	|
| Single-time (1) 	|           	|         	|           	|         	|
|    Before (2)   	| 115.17 ms 	| 4.57 ms 	| 138.76 ms 	| 4.98 ms 	|
|  Multi-time (2) 	| 122.30 ms 	| 4.02 ms 	| 146.28 ms 	| 4.25 ms 	|
| Single-time (2) 	|           	|         	|           	|         	|
|    Before (3)   	| 115.41 ms 	| 4.86 ms 	| 141.28 ms 	| 6.57 ms 	|
|  Multi-time (3) 	| 122.91 ms 	| 3.97 ms 	| 146.72 ms 	| 5.71 ms 	|
| Single-time (3) 	|           	|         	|           	|         	|

**NC** refers to non-contiguous.
**Multi-time** refers to the implementation with multi-time reshape and sum.
**Single-time** refers to the implementation with single-time reshape and sum.

ejguan committed:
Update on "Optimize backward for torch.repeat"


Fixes #43192

Differential Revision: [D24481801](https://our.internmc.facebook.com/intern/diff/D24481801)

[ghstack-poisoned]

ejguan commented:
**Benchmark for repeat/repeat_interleave/expand:**
```python
x = torch.rand((1, 1920)).requires_grad_()

repeated = x.repeat(1280, 1)
repeated = x.repeat_interleave(1280, dim=0)
repeated = x.expand(1280, 1920)
```
Average time for 10,000 runs:
|     | previous repeat |   repeat  | repeat_interleave |   expand  |
|:---:|:---------------:|:---------:|:-----------------:|:---------:|
| CPU |     6.71 ms     |  2.12 ms  |      1.45 ms      |  1.34 ms  |
| GPU |     15.72 ms    | 235.48 us |     564.45 us     | 168.74 us |

zou3519 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -662,14 +652,60 @@ Tensor repeat_backward(Tensor grad, IntArrayRef repeats, IntArrayRef input_shape
   }
-  for (size_t j = num_unsqueezed; j < repeats.size(); ++j) {
+
+  const auto & grad_sizes = grad.sizes();
```
nit: there's no need to do a reference for grad.sizes(); grad.sizes() returns an IntArrayRef which is a reference to the underlying buffer already

zou3519 replied:
Actually, this is UB: grad.sizes() returns an IntArrayRef, but we're assigning it to a reference so there's nothing that is actually owning the IntArrayRef to keep it alive

```

ejguan replied:
I think without `&`, `auto` is going to do copy construction from `IntArrayRef`. `auto &` is actually reference to the `IntArray`

zou3519 commented:
> Apparently, both of new implementations have better performance.
One-time strategy has slightly better performance, especially on CPU.

Makes sense for the one-time strategy to have slightly better performance; the multi-time strategy creates a lot of intermediate Tensors and incurs dispatching overhead every time an operation is called

zou3519 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -662,14 +652,60 @@ Tensor repeat_backward(Tensor grad, IntArrayRef repeats, IntArrayRef input_shape
+    //  [g2_2+g2_4, g2_3+g2_5]]                                     [g2_2, g2_3],
+    //                                   [[g2_2, g2_3],             [g2_4, g2_5]]
+    //                                    [g2_4, g2_5]]]
+    sum_dims.push_back(dim + offset);
+    ++offset;
+    grad_iter_ = grad_sizes_.insert(grad_iter_, repeat);
+    *(++grad_iter_) /= repeat;
+  }
```
We should give a concrete example of what grad_sizes_ and sum_dims comes out to be. For example, for input
```
x = torch.randn(2, 3, 5)
y = x.repeat(7, 1, 11)
```
grad_sizes is: [7, 2, 3, 11, 5]
sum_dims is: [0, 3] (e.g., we sum over all of the "repeat" dims)

ejguan replied:
make sense!

zou3519 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -662,14 +652,60 @@ Tensor repeat_backward(Tensor grad, IntArrayRef repeats, IntArrayRef input_shape
+    *(++grad_iter_) /= repeat;
+  }
+  grad = grad.reshape(grad_sizes_);
+  grad = grad.sum(sum_dims, false);
```
nit: the `false` is redundant, right?

ejguan replied:
Yep, default is false.

zou3519 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -662,14 +652,60 @@ Tensor repeat_backward(Tensor grad, IntArrayRef repeats, IntArrayRef input_shape
+    //  [g2_2+g2_4, g2_3+g2_5]]                                     [g2_2, g2_3],
+    //                                   [[g2_2, g2_3],             [g2_4, g2_5]]
+    //                                    [g2_4, g2_5]]]
+    sum_dims.push_back(dim + offset);
+    ++offset;
+    grad_iter_ = grad_sizes_.insert(grad_iter_, repeat);
+    *(++grad_iter_) /= repeat;
```
It might be easier to rewrite the iteration in terms of `input_shape` and `repeats`. That way, we wouldn't have to do division. In python pseudocode:
```
repeats = repeats[num_unsqueezed:]
grad_shape = []
sum_dims = []
for repeat, size in zip(repeats, input_shape):
    if repeat == 1:
        grad_shape.append(size)
        continue
    sum_dims.push_back(grad_shape.size())
    grad_shape += [repeat, size]
```
The benefits of iterating like this are:
- you don't have to keep around an offset
- no need to do division
- no need to "insert" elements into the grad_shape vector

ejguan replied:
Super, I was stuck with the previous logic. Nice.

zou3519 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -662,14 +652,60 @@ Tensor repeat_backward(Tensor grad, IntArrayRef repeats, IntArrayRef input_shape
+  std::copy(grad_sizes.begin(), grad_sizes.end(), grad_sizes_.begin());
+  auto grad_iter_ = grad_sizes_.begin();
+  int64_t offset = 0;
+  for (size_t j = num_unsqueezed; j < repeats.size(); ++j, ++grad_iter_) {
```
nit: `repeat_idx` instead of `j` is more descriptive

zou3519 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -662,14 +652,60 @@ Tensor repeat_backward(Tensor grad, IntArrayRef repeats, IntArrayRef input_shape
+    //                                    [g2_4, g2_5]]]
+    sum_dims.push_back(dim + offset);
+    ++offset;
+    grad_iter_ = grad_sizes_.insert(grad_iter_, repeat);
```
We can't do the algorithm this way. Inserting into the SmallVector may invalidate the iterator, just like how inserting into a std::vector can invalidate the iterator.

Inserting a value into a std::vector could cause the vector to reallocate memory and relocate its memory to that new region of memory. If this happens, then the next value of the original iterator is going to point to some deallocated memory and dereferencing it is UB

ejguan replied:
You are right, it's a bug. It may violate the memory location.

zou3519 reviewed (commented):
I have some suggestions on reframing the algorithm in terms of `input_shape` and `repeats` instead of using `grad_shape` and `repeats` that may make the logic easier to implement. Aside from that, I think there are some UBs in the code as-is, let me know what you think

ejguan committed:
Update on "Optimize backward for torch.repeat"


Fixes #43192

Differential Revision: [D24481801](https://our.internmc.facebook.com/intern/diff/D24481801)

[ghstack-poisoned]

ejguan commented:
> I have some suggestions on reframing the algorithm in terms of `input_shape` and `repeats` instead of using `grad_shape` and `repeats` that may make the logic easier to implement. Aside from that, I think there are some UBs in the code as-is, let me know what you think

Updated PR and the benchmark is updated.

zou3519 reviewed (approved):
lgtm!

ejguan committed:
Update on "Optimize backward for torch.repeat"


Fixes #43192

Differential Revision: [D24481801](https://our.internmc.facebook.com/intern/diff/D24481801)

[ghstack-poisoned]

ejguan committed:
Update on "Optimize backward for torch.repeat"


Fixes #43192

Differential Revision: [D24481801](https://our.internmc.facebook.com/intern/diff/D24481801)

[ghstack-poisoned]

ejguan committed:
Update on "Optimize backward for torch.repeat"


Fixes #43192

Differential Revision: [D24481801](https://our.internmc.facebook.com/intern/diff/D24481801)

[ghstack-poisoned]

ejguan commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -662,14 +652,65 @@ Tensor repeat_backward(Tensor grad, IntArrayRef repeats, IntArrayRef input_shape
+  // grad_size          [4, 2,    3, 9, 4, 3, 5]
+  // sum_dims           [0,          3,    5]
+  grad = grad.reshape(grad_size);
+  grad = grad.sum(sum_dims);
```
When we do repeat 1 at all dimensions, `sum_dims` becomes empty and leads to sum over the whole `grad` rather than keep it there.

zou3519 replied:
#29137 strikes again

ejguan committed:
Update on "Optimize backward for torch.repeat"


Fixes #43192

Differential Revision: [D24481801](https://our.internmc.facebook.com/intern/diff/D24481801)

[ghstack-poisoned]

zou3519 commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -709,8 +709,13 @@ Tensor repeat_backward(Tensor grad, IntArrayRef repeats, IntArrayRef input_shape
+  if (!sum_dims.empty()) {
+    grad = grad.reshape(grad_size);
+    grad = grad.sum(sum_dims);
+  }
```
Can we add a test for this?

ejguan replied:
Yes. Plan to do.

ejguan committed:
Update on "Optimize backward for torch.repeat"


Fixes #43192

Differential Revision: [D24739840](https://our.internmc.facebook.com/intern/diff/D24739840)

[ghstack-poisoned]

zou3519 commented:
If "import to phabricator" or ghimport fails, you might have to unlink this github PR from the original diff.
